design/one-pager-ignore-changes.md

```diff
@@ -0,0 +1,199 @@
+
+The implementation would be done on the provider level, so we would need to update all the providers over time to use the 
+new field. 
+
+For providers using Upjet the solution will be to leverage Terraform's [ignore_changes] lifecycle field. So Upjet
+generation should be updated with a step that transfers the `spec.ignoreChanges` field into Terraform's `ignore_changes`.
+Something like this in the Upjet method that writes the main TF file:
+
+```go
+// WriteMainTF writes the content main configuration file that has the desired
+// state configuration for Terraform.
+func (fp *FileProducer) WriteMainTF() error {
+
+    fp.parameters["lifecycle"] = map[string]interface{}{
+        // If the resource is in a deletion process, we need to remove the deletion
+        // protection.
+        "prevent_destroy": !meta.WasDeleted(fp.Resource),
+        // Add fields which should be ignored on updates.
+        "ignore_changes": fp.Resource.GetIgnored(),
+    }
+...
+```
+
+For other providers we can implement a helper function in the [crossplane-runtime] repo that would use the existing code
+in [crossplane-runtime/fieldpath] to unset the fields of a managed resource that are under `ignoreChanges`. That way
+provider maintainers could use this function in the Update method without the need to invent their own solution.
+
+Helper function would look something this (WIP):
+```go
+func UnsetIgnored(managed resource.Managed) error {
+	p, err := PaveObject(managed)
+	if err != nil {
+		return err
+	}
+	
+	for _, s := range managed.GetIgnored() {
+		err = p.DeleteField(s)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+```
+
+#### Feature Gating
+
+Similar to all other new features being added to Crossplane, we will ship this
+new policy as an alpha feature that will be off by default and will be
+controlled by the `--enable-alpha-ignore-changes` flag in Providers.
+
+This will not prevent the field from appearing in the schema of the managed
+resources. However, we will ignore the `spec.ignoreChanges` when the feature
+is not enabled.
+
+### Implementation alternative considered
+
+An alternative was considered where the above helper function which unsets fields in the managed resource would be used
+directly in the [crossplane-runtime] reconciler, and then a version of the managed resource without the ignored fields
+would be passed to the Update method.
+
+While this solution would enable the functionality without the need to change all the providers, it has a few 
+drawbacks:
+* It doesn't allow for flexibility or handling special cases in the provider code, as the providers will just receive a 
+managed resource which does not have the ignored fields set. There could be cases where the wish is to not update the field
+on the external resource, but is useful in some other way.
+* Won't work well with Upjet as it uses Terraform underneath which would work better with its native [ignore_changes].
+
+## Solution alternative considered
```

Comment:
I think another solution to consider could be to handle this problem on case by case basis depending on how many cases we've seen. For example, `NodeGroup` could have `initialSize` along `desiredSize` where both are marked as optional instead of a generic solution and we can manage it in the code.

In my experience, it's quite seldom in Crossplane land compared to Terraform since `ignore_changes` is used way more frequently in Terraform especially to prevent resource deletion & recreation since that's the behavior of TF for fields that cannot be changed whereas we simply do nothing or reject the action via webhook. Hence comparison with how many times `ignore_changes` is seen in HCLs or how it's top-level field in Terraform may be misleading. Ideally, we would gather a ratio like `every 1 of 4 CRDs need this functionality` (I suspect it's way lower) and then decide but that's probably quite challenging to gather other than trying to think of counting the open issues. Maybe one way could be to implement this field as an optional annotation and survey users about how many CRDs they needed to use it in. This could also be a good way to ship it as alpha without a change in CRDs, which is similar to how upstream k8s usually introduces new fields.

Reply:
The main use case I have seen for this functionality is with tagging. Many cloud users have external systems that add tags to resources for billing, tracking or similar purposes.